### ISSUE TITLE
Poor performance in EventSource manifest generation 

### ISSUE THREAD

**Issue by dmetzgar** [dotnet/coreclr#1461]
When an EventSource is first instantiated the manifest can be generated. During the manifest generation, a method called DebugCheckEvent is called on every event. One of the parameters is an array `EventMetaData[] eventData`. This array is constructed by finding the highest event id in the EventSource and allocating an array of that size. To maintain consistency between platforms, some libraries will need to keep the event ids from their old ETW providers (e.g. WCF). Some of the event ids may be very large. For example, WCF's highest event id is over 62K and only has 400 events. This creates a mostly empty array. Besides the waste of memory, there is the following code in EventSource:
```c#
for (int idx = 0; idx < eventData.Length; ++idx)
{
   if (eventData[idx].Descriptor.Task == (int)eventAttribute.Task && eventData[idx].Descriptor.Opcode == (int)eventAttribute.Opcode)
    {
        manifest.ManifestError(Environment.GetResourceString("EventSource_TaskOpcodePairReused",
                                evtName, evtId, eventData[idx].Name, idx));
    }
}
```
For each event, it will loop through the entire `eventData` array to determine if the task/opcode pair has been reused. This is inefficient with a mostly empty 62K array. Consider using a Dictionary or other data structure for EventSources that have large event ids compared to the number of actual events.

**Comment by jkotas** [comment]
cc @vancem

**Comment by vancem** [comment]
This is a known issue.   The goal of the eventData array is to allow very fast lookup at WriteEvent() time.   Changing the data structure to a dictionary looses some of the value of using numbers instead of names for event IDs (you might as well use strings, which are significantly more friendly).      Trading this speed of (which affects EVERY user of EventSource), for rare case of users who have constraints on their eventIDs seems like the wrong direction.  

Your situation seems like it should be exceedingly rare.   First it implies that you have a non-EventSource set of events, and you wanted to upgrade them to EventSource AND you were unwilling to break compatibility (at least on some of the events).     I believe that WCF is an instance, but wonder there is more than even 1 more case.    Yes, a 62K array is unfortunate, it is not insane number,  it is only in rare cases such as yours and you can fix it by renumbering over time (update tools to accept both and old and new number) and eventually abandon the old numbers.

**Comment by roncain** [comment]
@vancem , @dmetzgar , @mconnew -- we need to find a solution to this issue.  I don't think renumbering WCF event ID's is a tractable solution, given that it will break tooling and probably require changes to the desktop.  Something is unique in the TFS ToF environment that causes this to occur in manifest generation in CHK builds.  I've opened https://github.com/dotnet/wcf/issues/313 for the WCF team to investigate and propose a better solution.

**Comment by vancem** [comment]
My understanding of the code is that the DebugCheckEvent is NOT called during EventSource enablement, but only when you call GenerateManifest with the 'Strict' bit set on the flags parameter.   

If so you can avoid the CPU time simply by not passing 'Strict'. to that API.   

I have added a code to quickly skip any empty events which will also fix the problem but getting updates to EventSource is typically problematic since MSCORLIB needs to be updated.  .  

Note that neither of these mitigations reduce the size issue.   Your options really are to either live with it or renumber to avoid the wasted space.   Note that renumbering may not be as bad as you think depending on how many tools use the information (it is not hard to make code that would accept both new and old numbers).    That choice is of course up to you.

**Comment by dmetzgar** [comment]
For the 'Strict' bit I'm not sure how to not set that. I don't explicitly call GenerateManifest anywhere. I just have an EventSource class that creates a singleton instance. The (backwards) call stack looks like this:
- ManagedModule!Environment.GetResourceString                                                                                                       
- ManagedModule!EventSource.DebugCheckEvent                                                                                                        
- ManagedModule!EventSource.CreateManifestAndDescriptors                                                                                          
- ManagedModule!EventSource.EnsureDescriptorsInitialized                                                                                         
- ManagedModule!EventSource.DoCommand                                                                                                           
- ManagedModule!EventSource.Initialize                                                                                                         
- ManagedModule!System.Diagnostics.Tracing.EventSource..ctor(value class System.Diagnostics.Tracing.EventSourceSettings,class System.String[])
- ManagedModule!System.Diagnostics.Tracing.EventSource..ctor(value class System.Diagnostics.Tracing.EventSourceSettings)                     
- ManagedModule!System.Diagnostics.Tracing.EventSource..ctor()                                                                              
- ManagedModule!System.Runtime.WcfEventSource..ctor()                                                                                      
- ManagedModule!System.Runtime.WcfEventSource..cctor()                                                                                    

Note that this only happens on chk builds.

The other thing I noticed is regarding the code itself. I *assumed* from the DebugCheckEvent code that eventData is the size of the highest event id. But that's incorrect. Here is the code that creates the eventData array:
```c#
MethodInfo[] methods = eventSourceType.GetMethods(BindingFlags.DeclaredOnly | BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance);
EventAttribute defaultEventAttribute;
int eventId = 1;        // The number given to an event that does not have a explicitly given ID. 
EventMetadata[] eventData = null;
Dictionary<string, string> eventsByName = null;
if (source != null || (flags & EventManifestOptions.Strict) != 0)
{
    eventData = new EventMetadata[methods.Length + 1];
    eventData[0].Name = "";         // Event 0 is the 'write messages string' event, and has an empty name.
}
```
What this looks like is that it creates eventData to be the size of the number of instance methods found in the EventSource class. We have 461 events in the WcfEventSource. However, each event has an *IsEnabled event along with it. So eventData should be at least 923 elements in size.

The next thing that bugs me is that in our hot path above, the Environment.GetResourceString is being called. It should only call this if there's an error. I think that's coming from this code:
```c#
for (int idx = 0; idx < eventData.Length; ++idx)
{
    if (eventData[idx].Descriptor.Task == (int)eventAttribute.Task && eventData[idx].Descriptor.Opcode == (int)eventAttribute.Opcode)
    {
        manifest.ManifestError(Environment.GetResourceString("EventSource_TaskOpcodePairReused",
                                evtName, evtId, eventData[idx].Name, idx));
    }
}
```
What I think is happening is that since eventData is a struct and so is the Descriptor, that it defaults Task and Opcode to zero. If I also have an event that does not explicitly set the task and opcode, then they should also be zero (EventTask.None and EventOpcode.Info) and trigger the TaskOpcodePairReused manifest error. This would happen less than 923 x 461 times (425,503).

I think the only change that **we** would need is:
```c#
if (eventAttribute.Task != EventTask.None)
{
    for (int idx = 0; idx < eventData.Length; ++idx)
    {
        if (eventData[idx].Descriptor.Task == (int)eventAttribute.Task && eventData[idx].Descriptor.Opcode == (int)eventAttribute.Opcode)
        {
            manifest.ManifestError(Environment.GetResourceString("EventSource_TaskOpcodePairReused",
                                    evtName, evtId, eventData[idx].Name, idx));
        }
    }
}
```
But then there is the issue of eventData itself. Which should actually be fixed to be the size of the highest event id and not the count of the number of methods in the EventSource.

**Comment by vancem** [comment]
David,

Can you please open a bug on this that we should fix for the V4.6.1 release?



The bottom line here is that we don’t want this debug checking logic firing if people have gaps in their EventIds that they consume (we should have a test with exceptions turned on that confirm that gaps don’t cause swallowed errors).



I believe the actual fix is very easy new code in red in DebugCheck.    However we should be sure to confirm that it fixes their particular example



            // We give a task to things if they don't have one.
            Debug.Assert(eventAttribute.Task != EventTask.None || eventAttribute.Opcode != EventOpcode.Info);
            for (int idx = 0; idx < eventData.Length; ++idx)
            {
                // skip unused Event IDs.
                if (eventData[idx].Name == null)
                    continue;

                if (eventData[idx].Descriptor.Task == (int)eventAttribute.Task && eventData[idx].Descriptor.Opcode == (int)eventAttribute.Opcode)
                {
                    manifest.ManifestError(Environment.GetResourceString("EventSource_TaskOpcodePairReused",
                                            evtName, evtId, eventData[idx].Name, idx));
                }
            }


 Thanks


Vance

**Comment by vancem** [comment]
I see that my reading of the code was incorrect and I do see that we do this checking on every time the EventSource is activated.    

As mentioned we do have a fix that will  quickly skip empty slots (I have asked David Mason to follow up).   

Note however, I still believe there is a mystery here.   In my reading of code this checking logic should only be called on actual events (not blank slots)  and earlier logic looks for both Task and Opcode being 0 and sets the Task to something nonzero.  The upshot of this is that I don't see how the 'ManifestError call can happen by comparing against a EMTPY slot.    

I would like us to confirm or refute my reading of the code.   We can do this simply by looking at the errors that come back (it will give us the event Id of the two events involved and we can see if they are blank event or not.

I also can't explain why it only happens on CHK builds.  There is no CHK only logic here.   

I am suspicious that this is a 'real' error in the sense that in your CHK builds you have something that does indeed have a task and opcode that are the same.   There is a simmiar bug we ran into where earlier logic that auto-assigns Start and Stop opcodes causes this logic to fire.  

So can you tell use what event IDs it is complaining about?

P.S:  The array actually is as large as the largest event number.  It does start out being just as big as the number of methods (as you discovered), but there is logic in AddEventDescriptor that will extend the array if needed.  (which by the way is reasonably inefficient in your case.  We really are not expecting users to create large blank regions, so we only extend the array 16 entries at a time.  If you are resolved to keep your inefficient event numbering you should at least put the largest ID first, so that it only reallocates once).

**Comment by dmetzgar** [comment]
Here's the smallest repro of the error that I can come up with: https://github.com/dmetzgar/eventsourcetest
This is the code:
```c#
    public class Program
    {
        public void Main(string[] args)
        {
            try
            {
                string manifest = EventSource.GenerateManifest(typeof(SampleEventSource), "SomeAssembly.dll", EventManifestOptions.Strict);
                Console.WriteLine(manifest);
            }
            catch (Exception exc)
            {
                Console.WriteLine(exc.Message);
            }
            Console.WriteLine("Done.");
            Console.ReadLine();
        }
    }
    public sealed class SampleEventSource : EventSource
    {
        [Event(2, Level = EventLevel.Warning, Opcode = EventOpcode.Info,
            Message = "Throwing an exception.")]
        public void ThrowingEtwException()
        {
            WriteEvent(2);
        }
    }
```
It produces the following output:
```
Event ThrowingEtwException (with ID 2) has the same task/opcode pair as event  (with ID 0).
Event ThrowingEtwException (with ID 2) has the same task/opcode pair as event  (with ID 1).
Done.
```

**Comment by vancem** [comment]
David,

This is likely caused by logic that we added when we added the Activity support.     We certainly should not be logging an error in the case below, so we should add this to the bug you created.

Vance

**Comment by davmason** [comment]
This was fixed with efa1dda2c1579417dcaed92633b8556fd68270ee, it is also in desktop by now.
